raster/r.neighbors/readweights.c

```diff
@@ -25,20 +26,24 @@ void read_weights(const char *filename)
     fclose(fp);
 }
 
-void gaussian_weights(double sigma)
+void compute_weights(const char * function_type, double factor)
 {
-    double sigma2 = sigma * sigma;
-    int i, j;
-
-    ncb.weights = G_malloc(ncb.nsize * sizeof(DCELL *));
-    for (i = 0; i < ncb.nsize; i++)
-	ncb.weights[i] = G_malloc(ncb.nsize * sizeof(DCELL));
-
-    for (i = 0; i < ncb.nsize; i++) {
-	double y = i - ncb.dist;
-	for (j = 0; j < ncb.nsize; j++) {
-	    double x = j - ncb.dist;
-	    ncb.weights[i][j] = exp(-(x*x+y*y)/(2*sigma2))/(2*M_PI*sigma2);
+		int i, j;
+
+        ncb.weights = G_malloc(ncb.nsize * sizeof(DCELL *));
+        for (i = 0; i < ncb.nsize; i++)
+		ncb.weights[i] = G_malloc(ncb.nsize * sizeof(DCELL));
+
+        for (i = 0; i < ncb.nsize; i++) {
+		double y = i - ncb.dist;
+		for (j = 0; j < ncb.nsize; j++) {
+	        double x = j - ncb.dist;
+            if (strcmp(function_type, "gaussian") == 0) {
```

Comment:
Please use separate loops of `for (i = 0; i < ncb.nsize; i++) { ... }` for separate functions to avoid if statements in a nested loop. This is general code optimization and processing speed-up.

Reply:
Would I be on the right track to try to implement a function pointer before the loop that points to the chosen weighting function, which then gets executed within the loop? That would remove the if statement, right?

Reply:
Sounds right to me. Detect the function before the loop and then apply it in the loop.